llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -3214,6 +3238,107 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+  //   FalseBB:
+  //     store i32 1, ptr %q, align 4
+  //     ...
+  //   TrueBB:
+  //     %maskedloadstore = load i32, ptr %b, align 4
+  //     store i32 %maskedloadstore, ptr %p, align 4
+  //     ...
+  // \endcode
+  //
+  // and transform it into:
+  //
+  // \code
+  //   BB:
+  //     ...
+  //     %cond = icmp ult %x, %y
+  //     %maskedloadstore = cload i32, ptr %b, %cond
+  //     cstore i32 %maskedloadstore, ptr %p, %cond
+  //     cstore i32 1, ptr %q, ~%cond
+  //     br i1 %cond, label %TrueBB, label %FalseBB
+  //   FalseBB:
+  //     ...
+  //   TrueBB:
+  //     ...
+  // \endcode
+  //
+  // where cload/cstore are represented by llvm.masked.load/store intrinsics,
+  // e.g.
+  //
+  // \code
+  //   %vcond = bitcast i1 %cond to <1 x i1>
+  //   %v0 = call <1 x i32> @llvm.masked.load.v1i32.p0
+  //                         (ptr %b, i32 4, <1 x i1> %vcond, <1 x i32> poison)
+  //   %maskedloadstore = bitcast <1 x i32> %v0 to i32
+  //   call void @llvm.masked.store.v1i32.p0
+  //                          (<1 x i32> %v0, ptr %p, i32 4, <1 x i1> %vcond)
+  //   %cond.not = xor i1 %cond, true
+  //   %vcond.not = bitcast i1 %cond.not to <1 x i>
+  //   call void @llvm.masked.store.v1i32.p0
+  //              (<1 x i32> <i32 1>, ptr %q, i32 4, <1x i1> %vcond.not)
+  // \endcode
+  //
+  // So we need to turn hoisted load/store into cload/cstore.
+  auto &Context = BI->getParent()->getContext();
+  auto *VCondTy = FixedVectorType::get(Type::getInt1Ty(Context), 1);
+  auto *Cond = BI->getOperand(0);
+  Value *Mask = nullptr;
+  // Construct the condition if needed.
+  if (!SpeculatedConditionalLoadsStores.empty()) {
+    IRBuilder<> Builder(SpeculatedConditionalLoadsStores.back());
+    if (Invert)
+      Mask = Builder.CreateBitCast(
+          Builder.CreateXor(Cond, ConstantInt::getTrue(Context)), VCondTy);
+    else
+      Mask = Builder.CreateBitCast(Cond, VCondTy);
+  }
+  for (auto *I : SpeculatedConditionalLoadsStores) {
+    IRBuilder<> Builder(I);
+    // We currently assume conditional faulting load/store is supported for
+    // scalar types only when creating new instructions. This can be easily
+    // extended for vector types in the future.
+    assert(!getLoadStoreType(I)->isVectorTy() && "not implemented");
+    auto *Op0 = I->getOperand(0);
+    Instruction *MaskedLoadStore = nullptr;
+    if (auto *LI = dyn_cast<LoadInst>(I)) {
+      // Handle Load.
+      auto *Ty = I->getType();
+      MaskedLoadStore = Builder.CreateMaskedLoad(FixedVectorType::get(Ty, 1),
+                                                 Op0, LI->getAlign(), Mask);
+      I->replaceAllUsesWith(Builder.CreateBitCast(MaskedLoadStore, Ty));
+    } else {
+      // Handle Store.
+      auto *StoredVal =
+          Builder.CreateBitCast(Op0, FixedVectorType::get(Op0->getType(), 1));
+      MaskedLoadStore = Builder.CreateMaskedStore(
+          StoredVal, I->getOperand(1), cast<StoreInst>(I)->getAlign(), Mask);
```

Comment:
Reproducer (extracted from opencv):
```
; bin/llc -mattr=+cf reduced.ll
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

define void @_ZN2cv5errorERKNS_9ExceptionE(i1 %0) personality ptr null {
  br i1 %0, label %2, label %3

2:                                                ; preds = %1
  store i32 0, ptr null, align 4294967296
  br label %3

3:                                                ; preds = %2, %1
  ret void
}
```
```
masked_store: alignment must be a power of 2
  call void @llvm.masked.store.v1i32.p0(<1 x i32> zeroinitializer, ptr null, i32 0, <1 x i1> %2)
in function _ZN2cv5errorERKNS_9ExceptionE
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/llc -mattr=+cf reduced.ll
1.      Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.      Running pass 'Module Verifier' on function '@_ZN2cv5errorERKNS_9ExceptionE'
 #0 0x00007ffff040fbc2 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x20fbc2)
 #1 0x00007ffff040ca8f llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x20ca8f)
 #2 0x00007ffff040cbd5 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fffefa42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007fffefa969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007fffefa969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007fffefa969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007fffefa42476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007fffefa287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007ffff026c5a5 llvm::report_fatal_error(llvm::Twine const&, bool) (.cold) ErrorHandling.cpp:0:0
#10 0x00007ffff031b4b0 unsigned long std::uniform_int_distribution<unsigned long>::operator()<std::random_device>(std::random_device&, std::uniform_int_distribution<unsigned long>::param_type const&) (.isra.0) ExponentialBackoff.cpp:0:0
#11 0x00007ffff09c00c7 (anonymous namespace)::VerifierLegacyPass::runOnFunction(llvm::Function&) Verifier.cpp:0:0
#12 0x00007ffff08fec48 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMCore.so.20.0git+0x2fec48)
#13 0x00007ffff08ff251 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMCore.so.20.0git+0x2ff251)
#14 0x00007ffff08fdf27 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMCore.so.20.0git+0x2fdf27)
#15 0x000055555557448e compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#16 0x0000555555567e05 main (bin/llc+0x13e05)
#17 0x00007fffefa29d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#18 0x00007fffefa29e40 call_init ./csu/../csu/libc-start.c:128:20
#19 0x00007fffefa29e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#20 0x0000555555568c45 _start (bin/llc+0x14c45)
Aborted (core dumped)
```

Reply:
It is not guaranteed to be simplified by InstCombine because `null_pointer_is_valid` may be set.

Reply:
I doubt whether it's a valid alignment ....

Reply:
https://github.com/llvm/llvm-project/pull/99257 It seems the alignment should be less than 2^32 (4294967296) @dtcxzyw

Reply:
> #99257 It seems the alignment should be less than 2^32 (4294967296) @dtcxzyw

It is generated by a transform pass :( BTW `opt -O3 -verify-each` doesn't complain about this.
cc @nikic

Reply:
Okay. I found this PR https://reviews.llvm.org/D110451.  The alignment check is at https://github.com/llvm/llvm-project/blob/main/llvm/lib/IR/Verifier.cpp#L4227.  It seems the `<=` instead of `<` is on purpose. I will update my PR.

Reply:
Done. Tested by `not_maximum_alignment`